File: src/main/java/no/novari/configuration/validation/constraints/ValueParsableAsType.java
package no.novari.configuration.validation.constraints;

import jakarta.validation.Constraint;
import jakarta.validation.Payload;
import java.lang.annotation.Retention;

import static java.lang.annotation.RetentionPolicy.RUNTIME;

@Retention(RUNTIME)
@Constraint(validatedBy = ValueParsableAsTypeValidator.class)
public @interface ValueParsableAsType {

    String VALUE_TYPE = "valueType";

    String message() default "contains value that is not parsable as type={" + VALUE_TYPE + "}";

    Class<?>[] groups() default {};

    Class<? extends Payload>[] payload() default {};

}

File: src/main/kotlin/no/novari/flyt/configuration/validation/constraints/ValueParsableAsType.kt
package no.novari.flyt.configuration.validation.constraints

import jakarta.validation.Constraint
import jakarta.validation.Payload
import kotlin.reflect.KClass

@Retention(AnnotationRetention.RUNTIME)
@Target(AnnotationTarget.CLASS)
@Constraint(validatedBy = [ValueParsableAsTypeValidator::class])
annotation class ValueParsableAsType(
    val message: String = "contains value that is not parsable as type={$VALUE_TYPE}",
    val groups: Array<KClass<*>> = [],
    val payload: Array<KClass<out Payload>> = [],
) {
    companion object {
        const val VALUE_TYPE = "valueType"
    }
}
